src/bin/winkernel/ioctl.c

```diff
@@ -0,0 +1,389 @@
+/*++
+
+    Copyright (c) Microsoft Corporation.
+    Licensed under the MIT License.
+
+Abstract:
+
+    IOCTL interface for the MsQuic.sys driver.
+
+--*/
+
+#include "quic_platform.h"
+#include "quic_trace.h"
+#include "msquic.h"
+#include "msquicp.h"
+
+#ifdef QUIC_CLOG
+#include "ioctl.c.clog.h"
+#endif
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+QuicLibrarySumPerfCountersExternal(
+    _Out_writes_bytes_(BufferLength) uint8_t* Buffer,
+    _In_ uint32_t BufferLength
+    );
+
+DECLARE_CONST_UNICODE_STRING(QuicIoCtlDeviceName, L"\\Device\\msquic");
+DECLARE_CONST_UNICODE_STRING(QuicIoCtlDeviceSymLink, L"\\DosDevices\\msquic");
+
+typedef struct QUIC_DEVICE_EXTENSION {
+    void* Reserved;
+} QUIC_DEVICE_EXTENSION;
+
+WDF_DECLARE_CONTEXT_TYPE_WITH_NAME(QUIC_DEVICE_EXTENSION, QuicIoCtlGetDeviceContext);
+
+typedef struct QUIC_DRIVER_CLIENT {
+    void* Reserved;
+} QUIC_DRIVER_CLIENT;
+
+WDF_DECLARE_CONTEXT_TYPE_WITH_NAME(QUIC_DRIVER_CLIENT, QuicIoCtlGetFileContext);
+
+EVT_WDF_IO_QUEUE_IO_DEVICE_CONTROL QuicIoCtlEvtIoDeviceControl;
+EVT_WDF_IO_QUEUE_IO_CANCELED_ON_QUEUE QuicIoCtlEvtIoQueueCanceled;
+EVT_WDF_REQUEST_CANCEL QuicIoCtlEvtIoCanceled;
+
+PAGEDX EVT_WDF_DEVICE_FILE_CREATE QuicIoCtlEvtFileCreate;
+PAGEDX EVT_WDF_FILE_CLOSE QuicIoCtlEvtFileClose;
+PAGEDX EVT_WDF_FILE_CLEANUP QuicIoCtlEvtFileCleanup;
+
+WDFDEVICE QuicIoCtlDevice;
+QUIC_DEVICE_EXTENSION* QuicIoCtlExtension;
+
+_No_competing_thread_
+INITCODE
+NTSTATUS
+QuicIoCtlInitialize(
+    _In_ WDFDRIVER Driver
+    )
+{
+    NTSTATUS Status = STATUS_SUCCESS;
+    PWDFDEVICE_INIT DeviceInit = NULL;
+    WDF_FILEOBJECT_CONFIG FileConfig;
+    WDF_OBJECT_ATTRIBUTES Attribs;
+    WDFDEVICE Device;
+    QUIC_DEVICE_EXTENSION* DeviceContext;
+    WDF_IO_QUEUE_CONFIG QueueConfig;
+    WDFQUEUE Queue;
+
+    DeviceInit =
+        WdfControlDeviceInitAllocate(
+            Driver,
+            &SDDL_DEVOBJ_SYS_ALL_ADM_ALL);
```

Comment:
Who should have access to query these performance counters?

Reply:
Perfmon doesn't force admin, so I think it needs to be all users?